src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -1556,12 +1556,22 @@ public static SocketError AcceptAsync(Socket socket, SafeCloseSocket handle, Saf
 
         internal static SocketError DisconnectAsync(Socket socket, SafeCloseSocket handle, bool reuseSocket, DisconnectOverlappedAsyncResult asyncResult)
         {
-            throw new PlatformNotSupportedException(SR.net_sockets_disconnect_notsupported);
+            SocketError socketError = Disconnect(socket, handle, reuseSocket);
+            if (socketError == SocketError.Success)
+            {
+                asyncResult.CompletionCallback(SocketError.Success);
+            }
+
+            return socketError;
         }
 
         internal static SocketError Disconnect(Socket socket, SafeCloseSocket handle, bool reuseSocket)
         {
-            throw new PlatformNotSupportedException(SR.net_sockets_disconnect_notsupported);
+            using (Socket newsocket = new Socket(socket.AddressFamily, socket.SocketType, socket.ProtocolType))
+            {
+                Interop.Error err = Interop.Sys.Dup2(newsocket.SafeHandle, socket.SafeHandle);
+                return err == Interop.Error.SUCCESS ? SocketError.Success : GetSocketErrorForErrorCode(err);
+            }
```

Comment:
This ends up having the same problem that Mono's connect functionality has in general: the new socket handle isn't configured with any options that may have been set on the original socket, either via Socket's API or via getting Socket's Handle and doing direct configuration on the file descriptor via one's own P/Invokes.  If we want to enable this, we'll need to incorporate something along the lines of what was done in https://github.com/dotnet/corefx/pull/16373.  It's possible most of that should come "for free", since the {Begin}Connect{Async} methods are now instrumented to call ReplaceHandleIfNecessaryAfterFailedConnect() which handles the creation of a new socket and the propagation (where possible) of settings to it, so it may be that little else is needed here other than closing the old handle; I've not tried, though.